File: MatrixMultiplicationCPP/MatrixMultiplicationCPP/MatrixMultiplication.cpp
//Parallel and Distributed Computing - Assignment (Todai 2014)
//By Nikolai Gladkov

#include "stdafx.h"
#include <omp.h>
#include <ctime>
#include <iostream>

const int m = 1000, n = 1000, q = 1000;
float matrixA[m][n], matrixB[n][q], matrixC[m][q];

int main()
{
	//Initialization of matrices A and B
	for (int i = 0; i < m; i++)
	{
		for (int j = 0; j < n; j++)
		{
			matrixA[i][j] = (float)rand() / RAND_MAX * 100; 
		}
	}
	for (int j = 0; j < n; j++)
	{
		for (int k = 0; k < q; k++)
		{
			matrixB[j][k] = (float)rand() / RAND_MAX * 100;
		}
	}

	//Sequential matrix multiplication
	const clock_t startSequential = clock();
	for (int i = 0; i < m; i++)
	{
		for (int k = 0; k < q; k++)
		{
			float sum = 0;
			for (int j = 0; j < n; j++)
			{
				sum += matrixA[i][j] * matrixB[j][k]; 
			}
			matrixC[i][k] = sum;
		}
	}
	const double timeSequential = static_cast<double>(clock() - startSequential) / CLOCKS_PER_SEC;
	std::cout << "Sequential loop: " << timeSequential << " seconds" << std::endl;
	memset(matrixC, 0, sizeof(matrixC));

	//Parallel matrix multiplication via OpenMP
	const clock_t startOMP = clock();
	int i, j, k;
	omp_set_num_threads(4);
#pragma omp parallel for shared (matrixA, matrixB, matrixC) private (i, j, k)
	for (i = 0; i < m; i++)
	{
		for (k = 0; k < q; k++)
		{
			float sum = 0;
			for (j = 0; j < n; j++)
			{
				sum += matrixA[i][j] * matrixB[j][k]; 
			}
			matrixC[i][k] = sum;
		}
	}
	const double timeOMP = static_cast<double>(clock() - startOMP) / CLOCKS_PER_SEC;
	std::cout << "Parallel loop: " << timeOMP << " seconds" << std::endl;
	std::cout << "Speedup - OpenMP to Sequential: " << timeSequential/timeOMP << std::endl;

	return 0;
}

File: MatrixMultiplicationCS/MatrixMultiplicationCS/MatrixMultiplication.cs
//Parallel and Distributed Computing - Assignment (Todai 2014)
//By Nikolai Gladkov

using System;
using System.Diagnostics;
using System.Threading.Tasks;

class Program
{
    static void Main(string[] args)
    {
        const int m = 5000, n = 5000, q = 5000;
        float[,] matrixA = new float[m, n];
        float[,] matrixB = new float[n, q];
        float[,] matrixC = new float[m, q];
        Stopwatch stopwatch = new Stopwatch();
        Random rand = new Random();

        //Initialization of matrices A and B
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                matrixA[i, j] = (float)rand.NextDouble() * 100;
            }
        }
        for (int j = 0; j < n; j++)
        {
            for (int k = 0; k < q; k++)
            {
                matrixB[j, k] = (float)rand.NextDouble() * 100;
            }
        }

        //Sequential matrix multiplication
        stopwatch.Start();
        for (int i = 0; i < m; i++)
        {
            for (int k = 0; k < q; k++)
            {
                float sum = 0;
                for (int j = 0; j < n; j++)
                {
                    sum += matrixA[i, j] * matrixB[j, k];
                }
                matrixC[i, k] = sum;
            }
        }
        stopwatch.Stop();
        float timeSequential = (float)stopwatch.ElapsedMilliseconds / 1000;
        Console.WriteLine("Sequential loop: {0} seconds", timeSequential);
        stopwatch.Reset();
        matrixC = new float[m, q];

        //Parallel matrix multiplication
        stopwatch.Start();
        ParallelOptions options = new ParallelOptions();
        options.MaxDegreeOfParallelism = 2;
        Parallel.For(0, m, options, i =>
        {
            for (int k = 0; k < q; k++)
            {
                float sum = 0;
                for (int j = 0; j < n; j++)
                {
                    sum += matrixA[i, j] * matrixB[j, k];
                }
                matrixC[i, k] = sum;
            }
        });
        stopwatch.Stop();
        float timeParallel = (float)stopwatch.ElapsedMilliseconds / 1000;
        Console.WriteLine("Parallel loop: {0} seconds", timeParallel);
        Console.WriteLine("Speedup - Parallel to Sequential: {0}", timeSequential / timeParallel);

        Console.ReadKey();
    }
}